cibuildwheel/logger.py

```diff
@@ -0,0 +1,192 @@
+    'macosx_x86_64': 'macOS x86_64',
+}
+
+
+class Logger:
+    fold_mode: str
+    colors_enabled: bool
+    active_build_identifier: Optional[str] = None
+    build_start_time: Optional[float] = None
+    step_start_time: Optional[float] = None
+    active_fold_group_name: Optional[str] = None
+
+    def __init__(self):
+        if 'AZURE_HTTP_USER_AGENT' in os.environ:
+            self.fold_mode = 'azure'
+            self.colors_enabled = True
+
+        elif 'GITHUB_ACTIONS' in os.environ:
+            self.fold_mode = 'github'
+            self.colors_enabled = True
+
+        elif 'TRAVIS' in os.environ:
+            self.fold_mode = 'travis'
+            self.colors_enabled = True
+
+        elif 'APPVEYOR' in os.environ:
+            self.fold_mode = 'disabled'
+            self.colors_enabled = True
+
+        else:
+            self.fold_mode = 'disabled'
+            self.colors_enabled = file_supports_color(sys.stdout)
+
+    def build_start(self, identifier: str):
+        c = self.colors
+        description = build_description_from_identifier(identifier)
+        print()
+        print(f'{c.bold}{c.blue}Building {identifier} wheel{c.end}')
+        print(f'{description}')
+        print()
+
+        self.build_start_time = time.time()
+        self.active_build_identifier = identifier
+
+    def build_end(self):
+        assert self.build_start_time is not None
+        assert self.active_build_identifier is not None
+        self.step_end()
+
+        c = self.colors
+        duration = time.time() - self.build_start_time
+
+        print()
+        print(f'{c.green}✓ {c.end}{self.active_build_identifier} finished in {duration:.2f}s')
+        self.build_start_time = None
+        self.active_build_identifier = None
+
+    def step(self, step_description: str):
+        self.step_end()
+        self.step_start_time = time.time()
+        self._start_fold_group(step_description)
+
+    def step_end(self, success=True):
+        if self.step_start_time is not None:
+            self._end_fold_group()
+            c = self.colors
+            duration = time.time() - self.step_start_time
+            if success:
+                print(f'{c.green}✓ {c.end}{duration:.2f}s'.rjust(78))
+            else:
+                print(f'{c.red}✕ {c.end}{duration:.2f}s'.rjust(78))
+
+            self.step_start_time = None
+
+    def error(self, error: Union[Exception, str]):
+        self.step_end(success=False)
+        print()
+
+        if self.fold_mode == 'github':
+            print(f'::error::{error}')
+        else:
+            c = self.colors
+            print(f'{c.bright_red}Error{c.end} {error}')
+
+    def _start_fold_group(self, name: str):
+        self._end_fold_group()
+        self.active_fold_group_name = name
+        fold_start_pattern = FOLD_PATTERNS.get(self.fold_mode, DEFAULT_FOLD_PATTERN)[0]
+
+        print(fold_start_pattern.format(name=self.active_fold_group_name))
+        print()
```

Comment:
Just curious: why this extra empty line in the middle after the start of a group (and not after the end)?

Reply:
I guess I thought it looks nicer :) the newline appears inside the group, spaces it down a bit

e.g.
this:
![image](https://user-images.githubusercontent.com/1244307/97966425-bb150300-1db3-11eb-99b3-91858fa322ce.png)

versus this:
![image](https://user-images.githubusercontent.com/1244307/97966481-ce27d300-1db3-11eb-9117-c14a75c6d820.png)

dont feel strongly about it though!

Reply:
> Is the indentation automatically done by GHA? It looks kind of funky in combination with output from those commands:

No, the indentation is from our code. that's why the auditwheel stuff is not indented. Maybe there's a way to indent the subcommand output, but probably not worth the extra complexity.